src/transport/mod.rs

```diff
@@ -0,0 +1,150 @@
+use std::pin::Pin;
+
+use hyper::{body, Body, Request, Response};
+use wasm_bindgen::prelude::*;
+use wasm_bindgen::JsCast;
+use wasm_bindgen_futures::JsFuture;
+
+mod build;
+
+pub use build::TransportBuilder;
+
+use build::{HttpsConnector, HyperClient, WasmClient};
+
+pub struct Transport<C> {
+    client: C,
+}
+
+impl Transport<HyperClient<HttpsConnector>> {
+    pub fn new() -> Self {
+        TransportBuilder::new_hyper().build()
+    }
+}
+
+impl<C> Transport<HyperClient<C>>
+where
+    C: hyper::client::connect::Connect + Clone + Send + Sync + 'static,
+{
+    pub async fn send(
```

Comment:
send looks like something that could be in a Trait, 

```rust
use async_trait::async_trait;

/// http client
#[async_trait]
pub trait TransportTrait {
    /// method used by this client
    type Method;
    /// headers
    type HeaderMap;
    /// body type
    type Body;
    /// respone type
    type Response;

    /// send request!
    async fn send(
        &self,
        method: Self::Method,
        uri: &str,
        headers: Self::HeaderMap,
        body: Self::Body,
    ) -> Self::Response;
}
```

What do you think? And then we could implement that for each one of the Transport<C:> options that we have.

Reply:
That is a nice way to approach things.

Moreover, I was planning to replace `Method`, `HeaderMap`, and `Body` with a `Request`.

Reply:
Yes, that is a good option but also could be a `type Request` in pub trait HttpLowLevel, so we can use custom Request types from each libraries if they exist, I am not sure about that.

Reply:
```rust
#[async_trait]
pub trait HttpLowLevel {
    type Method;
    type HeaderMap;
    type Body;
    type Request;
    type Response;

    fn make_request(
        method: Self::Method,
        uri: &str,
        headers: Self::HeaderMap,
        body: Self::Body,
    ) -> Self::Request;

    async fn send(&self, request: Self::Request) -> Self::Response;
}
```

Is separating out the request making to a different function a good idea?

Reply:
Yes, that sounds fine, maybe that should be pub(crate) ??